src/render_tree.rs

```diff
@@ -0,0 +1,438 @@
+        self.x = relative_position.x + x_offset;
+        self.y = relative_position.y + y_offset;
+    }
+
+    /// Adjust y by an offset.
+    /// y += offset_y
+    pub fn offset_y(&mut self, offset_y: f64) {
+        self.y += offset_y;
+    }
+
+    /// Adjust x by an offset.
+    /// x += offset_x
+    pub fn offset_x(&mut self, offset_x: f64) {
+        self.x += offset_x;
+    }
+}
+
+impl Default for Position {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A RenderTree is a data structure to be consumed by a user agent
+/// that combines the DOM and CSSOM to compute layouts and styles
+/// for objects to draw on the screen.
+pub struct RenderTree {
+    /// Pointer to the underlying document that builds this render tree
+    // TODO: Add CSSOM here as well when we get to that point
+    document: DocumentHandle,
+    /// Entry point of the render tree
+    // TODO: make this a NodeHandle to make operations easier and not
+    // have to keep doing borrow(), borrow_mut() etc.
+    pub root: Rc<RefCell<Node>>,
+    /// Global render cursor position managed by the tree
+    /// that determines where to draw objects on the screen.
+    /// Whenever a node is created, the current value of the
+    /// position is copied into the node.
+    pub position: Position,
+}
+
+impl RenderTree {
+    pub fn new(document: &DocumentHandle) -> Self {
+        Self {
+            document: Document::clone(document),
+            root: Rc::new(RefCell::new(Node::new())),
+            position: Position::new(),
+        }
+    }
+
+    pub fn build(&mut self) {
+        // start with a clean root (if build is called multiple times)
+        self.root = Rc::new(RefCell::new(Node::new()));
+
+        let tree_iterator = document::TreeIterator::new(&self.document);
+        let mut reference_element = Rc::clone(&self.root);
+
+        for current_node_id in tree_iterator {
+            let doc_read = self.document.get();
+            if let Some(current_node) = doc_read.get_node_by_id(current_node_id) {
+                match &current_node.data {
+                    NodeData::Element(element) => {
+                        let new_node = match element.name.as_str() {
+                            "h1" => Node::new_heading1,
+                            "h2" => Node::new_heading2,
+                            "h3" => Node::new_heading3,
+                            "h4" => Node::new_heading4,
+                            "h5" => Node::new_heading5,
+                            "h6" => Node::new_heading6,
+                            "p" => Node::new_paragraph,
+                            _ => continue, /* add more here, for now ignore */
+                        }(&mut self.position);
+
+                        let new_node = Rc::new(RefCell::new(new_node));
+
+                        util::add_text_node(&reference_element, &new_node);
+                        reference_element = Rc::clone(&new_node);
```

Comment:
I just realized, the `util::add_text_node` is called here which is in the scope of the `NodeData::Element`. We will have to change this when we encounter different node types besides text (`span`, `div`, etc.)

Maybe we could update a `node_type` or something in the `new_node` match and call the appropriate util function for that node type.

For now this is fine, but will need update it when we get to that point

Reply:
Ah... Completely missed that one. Maybe we can return a tuple with the `Node` creation function and the util function to add the node

Reply:
Yeah, that could work. I'm not too worried about it at the moment since we only have one type right now and shouldn't be hard to extend it when we get to that point. I will probably just leave a note for now as a reminder. In the meantime I want to figure out the FFI stuff before I come back to the render tree (unless it's something you're interested in)

Reply:
ah, the function name is actually misleading. It has nothing to do with text nodes, it actually works for any generic node. I think I had that name in place before I abstracted it a bit, not sure. I'm just going to change the name and leave it how we have it